R/http-libcurl.R

```diff
@@ -87,75 +44,101 @@ httpLibCurl <- function(protocol,
       upload = TRUE,
       infilesize_large = fileLength,
       readfunction = function(nbytes, ...) {
-        if (is.null(con)) {
+        if (is.null(contentCon)) {
           return(raw())
         }
-        bin <- readBin(con, "raw", nbytes)
+        bin <- readBin(contentCon, "raw", nbytes)
         if (length(bin) < nbytes) {
-          close(con)
-          con <<- NULL
+          close(contentCon)
+          contentCon <<- NULL
         }
         bin
       }
     )
   }
 
-  # ensure we're using the requested method
-  curl::handle_setopt(handle, customrequest = method)
-
-  # apply all our accumulated headers
+  headers <- appendCookieHeaders(request, headers)
   curl::handle_setheaders(handle, .list = headers)
 
   # make the request
-  response <- NULL
+  url <- buildHttpUrl(request)
   time <- system.time(
     response <- curl::curl_fetch_memory(url, handle = handle),
     gcFirst = FALSE
   )
   httpTrace(method, path, time)
 
-  # get list of HTTP response headers
+  # Process headers
   headers <- curl::parse_headers_list(rawToChar(response$headers))
 
-  # read normalized location header
-  if ("location" %in% names(headers))
-    location <- headers[["location"]]
-  else
-    location <- NULL
+  # Parse cookies from header; bear in mind that there may be multiple headers
+  cookieHeaders <- headers[names(headers) == "set-cookie"]
+  storeCookies(request, cookieHeaders)
 
   # presume a plain text response unless specified otherwise
-  contentType <- if ("content-type" %in% names(headers)) {
-    headers[["content-type"]]
-  } else {
-    "text/plain"
-  }
+  contentType <- headers[["content-type"]] %||% "text/plain"
+  contentValue <- rawToChar(response$content)
 
   # emit JSON trace if requested
-  jsonTracingEnabled <-
-    httpTraceJson() &&
-    !is.null(contentFile) &&
-    identical(contentType, "application/json")
+  jsonTracingEnabled <- httpTraceJson() && contentType == "application/json"
+  if (jsonTracingEnabled) {
+    if (!is.null(contentFile)) {
+      cat(paste0("<< ", readLines(contentFile, warn = FALSE), "\n", collapse = ""))
+    }
+    lines <- strsplit(contentValue, "\n")[[1]]
+    cat(paste0(">> ", lines, "\n", collapse = ""))
+  }
 
-  if (jsonTracingEnabled)
-    cat(paste0("<< ", paste(readLines(contentFile), collapse = "\n"), "\n"))
+  list(
+    req = request,
+    status = response$status_code,
+    location = headers$location,
+    contentType = contentType,
+    content = contentValue
+  )
+}
 
-  # Parse cookies from header; bear in mind that there may be multiple headers
-  cookieHeaders <- headers[names(headers) == "set-cookie"]
-  storeCookies(list(protocol = protocol, host = host, port = port, path = path), cookieHeaders)
+createCurlHandle <- function(method,
```

Comment:
nit: Moving this function in a commit with functional code changes makes for poor diffs.